samples/runtime/reversible-simulator-advanced/simulator/Simulator.cs

```diff
@@ -0,0 +1,141 @@
+        public override void ControlledX(IQArray<Qubit> controls, Qubit qubit) {
+            if (controls.All(control => GetValue(control))) {
+                InvertValue(qubit);
+            }
+        }
+
+        // If the simulation value of qubit is true, when it's being reset, we
+        // restore it to false.
+        public override void Reset(Qubit qubit) {
+            SetValue(qubit, false);
+        }
+
+        // Measuring the qubit corresponds to translating the current simulation
+        // value into a Result value.
+        public override Result M(Qubit qubit) =>
+            GetValue(qubit).ToResult();
+
+        // Overriding the Assert methods enables the use of statements such as
+        // `AssertQubit(Zero, q)` inside a Q# program.
+        public override void Assert(IQArray<Pauli> bases, IQArray<Qubit> qubits, Result expected, string msg) {
+            Qubit filter(Pauli p, Qubit q) =>
+                p switch {
+                    Pauli.PauliI => null,
+                    Pauli.PauliZ => q,
+                    _ => throw new InvalidOperationException("Assert on bases other than PauliZ not supported")
+                };
+
+            // All qubits are considered whose corresponding measurement basis
+            // is PauliZ.  Measurement in PauliI basis are ignored, and other
+            // bases will raise an exception.
+            var qubitsToMeasure = bases.Zip(qubits, filter).WhereNotNull();
+
+            // A multi-qubit measurement in the PauliZ basis corresponds to
+            // computing the parity of all involved qubits' measurement values.
+            // (see also https://docs.microsoft.com/quantum/concepts/pauli-measurements#multiple-qubit-measurements)
+            // We use Aggregate to successively XOR a qubit's simulation value
+            // to an accumulator value `accu` that is initialized to `false`.
+            var actual = qubitsToMeasure.Aggregate(false, (accu, qubit) => accu ^ GetValue(qubit));
+            if (actual.ToResult() != expected) {
+                // If the expected value does not correspond to the actual measurement
+                // value, we throw a Q# specific Exception together with the user
+                // defined message `msg`.
+                throw new ExecutionFailException(msg);
+            }
+        }
+    }
+
+    public class ReversibleSimulator : QuantumProcessorDispatcher {
+        // Whether an exception is thrown when releasing qubits that are in a
+        // non-zero state is configured as an argument to the constructor of
+        // ReversibleSimulator, defaulted to `true`.
+        public ReversibleSimulator(bool throwOnReleasingQubitsNotInZeroState = true)
+            : base(new ReversibleSimulatorProcessor { ThrowOnReleasingQubitsNotInZeroState = throwOnReleasingQubitsNotInZeroState }) {
+
+            // We register the ApplyAnd and the ApplyLowDepthAnd operation as
```

Comment:
It might be helpful to clarify that these are both standard library operations.

Reply:
I added a comment pointing that out.